File: ts/src/base/ws/Stream.ts
import { sleep } from "../functions/time.js";
import { Int, Message, Topic, ConsumerFunction, BaseStream, Dictionary } from "../types";
import { Consumer } from "./Consumer.js";

export class Stream implements BaseStream {
    public maxMessagesPerTopic: Int;

    public topics: Dictionary<Message[]>;

    public verbose: boolean;

    public consumers: Dictionary<Consumer[]>;

    public topicIndexes: Dictionary<Int> = {};

    public activeWatchFunctions: any[] = [];

    constructor (maxMessagesPerTopic = 100, verbose = false) {
        this.init (maxMessagesPerTopic, verbose);
    }

    init (maxMessagesPerTopic = 10, verbose = false) {
        this.maxMessagesPerTopic = maxMessagesPerTopic
        this.verbose = verbose;
        this.topics = {};
        this.consumers = {};
        this.topicIndexes = {};
    }

    produce (topic: Topic, payload: any, error: any = null): void {
        if (!this.topics[topic]) {
            this.topics[topic] = [];
        }
        if (!(topic in this.topicIndexes)) {
            this.topicIndexes[topic] = -1;
        }

        this.topicIndexes[topic] += 1;
        const index = this.topicIndexes[topic];
        const messages = this.topics[topic];

        const message: Message = {
            payload,
            error,
            metadata: {
                stream: this,
                topic,
                index,
                history: messages.slice (),
            },
        };

        if (messages.length > this.maxMessagesPerTopic) {
            messages.shift ();
        }

        if (this.maxMessagesPerTopic !== 0) {
            this.topics[topic].push (message);
        }

        const consumers = this.consumers[topic] || [];
        this.sendToConsumers (consumers, message);
    }

    /**
     * Subscribes to a topic and registers a consumer function to handle incoming data.
     * Current index is the index of the last message consumed
     * @param topic - The topic to subscribe to.
     * @param consumerFn - The consumer function to handle incoming data.
     * @param synchronous - Optional. Indicates whether the consumer function should be executed synchronously or asynchronously. Default is true.
     */
    subscribe (topic: Topic, consumerFn: ConsumerFunction, synchronous: boolean = true): void {
        const consumer = new Consumer (consumerFn, synchronous, this.getLastIndex (topic));

        if (!this.consumers[topic]) {
            this.consumers[topic] = [];
        }

        this.consumers[topic].push (consumer);
        if (this.verbose) {
            console.log ('subscribed function to topic: ', topic, ' synchronous: ', synchronous);
        }
    }

    /**
     * Unsubscribes a consumer function from a topic.
     * @param topic - The topic to unsubscribe from.
     * @param consumerFn - The consumer function to unsubscribe.
     * @returns true if the consumer was successfully unsubscribed, false otherwise.
     */
    unsubscribe (topic: Topic, consumerFn: ConsumerFunction): boolean {
        if (this.consumers[topic]) {
            const consumersForTopic = this.consumers[topic];
            this.consumers[topic] = consumersForTopic.filter ((consumer) => consumer.fn !== consumerFn);
            if (this.verbose) {
                console.log ('unsubscribed function from topic: ', topic);
            }
            return true
        } else {
            if (this.verbose) {
                console.log ('unsubscribe failed: consumer not found for topic: ', topic);
            }
            return false
        }
    }

    /**
     * Retrieves the message history for a given topic.
     * @param topic - The topic for which to retrieve the message history.
     * @returns An array of messages representing the message history for the given topic.
     */
    getMessageHistory (topic: Topic): Message[] {
        return this.topics[topic] || [];
    }

    /**
     * Returns the last index of the given topic.
     * @param topic - The topic to get the last index for.
     * @returns The last index of the topic.
     */
    getLastIndex (topic: Topic): Int {
        if (topic in this.topicIndexes) {
            return this.topicIndexes[topic];
        }
        return -1;
    }

    /**
     * Sends a message to all consumers of a topic.
     * @param consumers {Consumer[]} - array of consumers
     * @param message {Message} - message to publish
     */
    private async sendToConsumers (consumers: Consumer[], message: Message): Promise<void> {
        if (this.verbose) {
            console.log ('sending message from topic ', message.metadata.topic, 'to ', consumers.length, ' consumers');
        }
        for (let i = 0; i < consumers.length; i++) {
            await Promise.resolve ();
            consumers[i].publish (message);
        }
        return;
    }

    /**
     * Adds a watch function to the stream. This is used to reconnect in case of any disconnectons.
     * @param watchFn {string} - the watchFunction name
     * @param args {any[]} - the arguments to pass to the watchFunction
     */
    public addWatchFunction (watchFn: string, args: any[]): void {
        this.activeWatchFunctions.push ({ 'method': watchFn, 'args': args })
    }

    /**
     * Closes the stream by resetting the topics and unsubscribing consumers
     * Note: this won't cancel any ongoing consumers
     */
    close (): void {
        if (this.verbose) {
            console.log ('closing stream');
        }
        this.init (this.maxMessagesPerTopic, this.verbose);
    }
}

export default Stream

File: python/ccxt/async_support/base/ws/stream.py
from typing import Any, Callable, Dict, List
from asyncio import Future
from ....base.types import Topic, Message
from .consumer import Consumer

# Define type aliases for clarity
MessagePayload = Any
ConsumerFunction = Callable[['Message'], Future[None] | None]


class Stream:
    def __init__(self, max_messages_per_topic: int = 10000, verbose: bool = False):
        """
        Initializes a new Stream object.
        :param int max_messages_per_topic: Maximum number of messages per topic. Defaults to None.
        """
        self.topics: Dict[Topic, List[Message]] = {}
        self.consumers: Dict[Topic, List[Consumer]] = {}
        self.max_messages_per_topic = max_messages_per_topic
        self.verbose = verbose
        self.active_watch_functions: List[str, List[Any]] = []
        if self.verbose:
            print('stream initialized')

    def produce(self, topic: Topic, payload: Any, error: Any = None) -> None:
        """
        Produces a new message for a topic.
        :param Topic topic: The topic to produce the message for.
        :param Any payload: The payload of the message.
        :param Any error: Any error associated with the message. Defaults to None.
        """
        if topic not in self.topics:
            self.topics[topic] = []

        messages = self.topics[topic]
        index = self.get_last_index (topic) + 1

        message = Message(payload, error, self, topic, index)

        if self.max_messages_per_topic and len(messages) >= self.max_messages_per_topic:
            messages.pop(0)

        self.topics[topic].append(message)
        consumers = self.consumers.get(topic, [])
        self.send_to_consumers(consumers, message)
        if self.verbose:
            print(f'Produced message for topic: {topic}. Total messages now: {len(messages)}.')

    def subscribe(self, topic: Topic, consumer_fn: ConsumerFunction, synchronous: bool = True) -> None:
        """
        Subscribes a consumer function to a topic.
        :param Topic topic: The topic to subscribe to.
        :param ConsumerFunction consumer_fn: The consumer function to subscribe.
        :param bool synchronous: Whether the consumer function should be called synchronously. Defaults to True.
        """
        consumer = Consumer(consumer_fn, synchronous, self.get_last_index(topic) + 1)
        if topic not in self.consumers:
            self.consumers[topic] = []
        self.consumers[topic].append(consumer)
        if self.verbose:
            print(f'Subscribed to topic: {topic}. Total consumers for topic now: {len(self.consumers[topic])}.')

    def unsubscribe(self, topic: Topic, consumer_fn: ConsumerFunction) -> None:
        """
        Unsubscribes a consumer function from a topic.
        :param Topic topic: The topic to unsubscribe from.
        :param ConsumerFunction consumer_fn: The consumer function to unsubscribe.
        """
        if topic in self.consumers:
            self.consumers[topic] = [consumer for consumer in self.consumers[topic] if consumer.fn != consumer_fn]
            if self.verbose:
                print(f'Unsubscribed consumer from topic: {topic}. Total consumers for topic now: {len(self.consumers[topic])}.')
        else:
            print(f'Unable to unsubscribe. Could not find consumer for topic: {topic}')

    def get_message_history(self, topic: Topic) -> List[Message]:
        """
        Gets the message history for a topic.
        :param Topic topic: The topic to get the message history for.
        :returns List[Message]: The message history for the topic.
        """
        return self.topics.get(topic, [])


    def get_last_index (self, topic: Topic) -> int:
        """
        Gets the last index for a topic. Returns -1 if the topic has no messages.
        :param Topic topic: The topic to get the last index for.
        :returns int: The last index for the topic.
        """
        last_index = -1
        messages = self.topics.get(topic, [])
        if len(messages) > 0:
            last_index = messages[-1].metadata.index
        return last_index

    def send_to_consumers(self, consumers, message):
        for consumer in consumers:
            consumer.publish(message)
        if self.verbose:
            print(f'Sending message to {len(consumers)} consumers for topic {message.metadata.topic}.')
    
    def add_watch_function (self, watchFn: str, args: List[Any]):
        self.active_watch_functions.append ({ 'method': watchFn, 'args': args })


    async def close(self) -> None:
        """
        Closes the stream. This cancels all pending tasks.
        """
        for consumers in self.consumers.values():
            for consumer in consumers:
                await consumer.close()
        if self.verbose:
            print("closed stream")
        self.stream = Stream()
